src/Shared/ITranslator.cs

```diff
@@ -16,6 +16,13 @@ namespace Microsoft.Build.BackEnd
     /// <typeparam name="T">The type to be translated.</typeparam>
     internal delegate T NodePacketValueFactory<T>(ITranslator translator);
 
+    /// <summary>
+    /// Delegate for users that want to translate an arbitrary structure that doesn't implement <see cref="ITranslatable"/> (e.g. translating a complex collection)
+    /// </summary>
+    /// <param name="translator">the translator</param>
+    /// <param name="objectToTranslate">the object to translate</param>
+    internal delegate void ObjectTranslator<T>(ITranslator translator, ref T objectToTranslate);
```

Comment:
Why not delete `NodePacketValueFactory` and replace it with `ObjectTranslator`?

Reply:
There's lots of code which already uses NodePacketValueFactory.  A type that implements ITranslatable either needs to have a default constructor or have something like NodePacketValueFactory.  I wasn't sure I wanted to commit to saying that all those types need to support default constructors.

Reply:
But they can just as easily provide an ObjectTranslator, instead of NodePacketValueFactory, right? ObjectTranslator seems to me to be able to completely replace NodePacketValueFactory.